src/module_1/module_1_meteo_api.py

```diff
@@ -14,16 +14,11 @@
 
 
 def main():
-    api_data = get_data_meteo_api("Madrid")
-    daily_data = api_data["daily"]
-    # a = [1, 3, 5, 7, 8]
-    b = get_data_meteo_api("Madrid")
-    c = b["daily"]["temperature_2m_mean"]
-    # print(c[:300])
-    print(mean_calculation(c))
-    print(variance_calculation(c, mean_calculation(c)))
-    plotting(daily_data)
-    # print(yearly_data)
+    for city in COORDINATES:
+        print(city)
+        api_data = get_data_meteo_api(city)
+        daily_data = api_data["daily"]
+        plotting(daily_data, city)
 
 
 def get_data_meteo_api(city):
```

Comment:
Comments on api_connection:

- What if the API rejects for too many calls? We could get the information we want just by waiting a little bit. That is much better than simply return an error to the user. We should be implementing here a retry policy that if we fail due to Errors, we sleep the process and retry in a while.

- requests.ConnectionError is the same as requests.exceptions.ConnectionError, so we have redundant code

- using Loggers is much more recommended that prints, since logs can be redirected to a file or any other place of interest easierly and with better management that just prints.

- The code is not managing the exceptions, nor reraising them, the user will get a None and it is unknown to him if that is the response or it has been an error. In addition, he will be unable to manage the different exceptions in his code flow

- Method is missing typehinting.

- The method is tied to this specific case when with little twists we could have a reusable method. For example, we could pass the url and any addition

Reply:
Comments on mean_calculation and variance_calculation:

 - Python is super slow on doing loops, always try to use vectorised operations when possible. You can use packages like numpy for that.
 - Mean and and std exist as built-in functions either in python standard library (native python) or using packages like numpy. These implementations are much faster since they are in C++ under the hood, giving a lot of efficiency to your code
 - Why rounding? Compute the exact values always unless there is a good reason. If the reason is "plotting" those are the **minor** changes once could expect to be done inside the plotting function itself.
 - Type hinting and docstrings again.

Reply:
Comments on daily to yearly:

- Type hinting and docstring
- When parsing dates use datetime package so you are sure you are parsing what you want. Instead of using .spit("-") you can use `datetime.strptime(date_string, format).year` with format `%Y-%m-%d`
- Following the try not to use loops, but vectorisation instead. You can reformat the code to use `pandas` as input data, extract the year of each date in vectorised form using pandas and then apply the "group by" command to apply the mean function to each year of data. Refer to [pandas intro documentation](https://pandas.pydata.org/pandas-docs/stable/user_guide/10min.html) if you are unfamiliar with the package.